File: test/scripts/trace_test.py
#
# This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
# Copyright (c) 2023-2024 Gabriele Digregorio, Roberto Alessandro Bertolini. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for details.
#

import unittest

from libdebug import debugger

class TraceTest(unittest.TestCase):
    def setUp(self):
        self.d = debugger("binaries/trace_test")

    def test_trace_basic(self):
        d = self.d
        d.run()
        bp = d.breakpoint(0x40114d)
        bbp = d.breakpoint(0x401141)
        d.cont()
        d.trace() #@ 4d
        d.finish(heuristic="step-mode")
        self.assertTrue(d._internal_debugger.trace_counter == 19)
        d.cont()
        self.assertTrue(d._internal_debugger.trace_counter == 38)
        index=1
        while index < 10:
            if index==4:
                d.step()
                self.assertTrue(d._internal_debugger.trace_counter == 96)
            elif index==7:
                self.assertTrue(d._internal_debugger.trace_counter == 152)
                d.step_until(0x401136)
                self.assertTrue(d._internal_debugger.trace_counter == 167)
            elif index==9:
                self.assertTrue(d._internal_debugger.trace_counter == 190)
            d.cont()
            index+=1
        self.assertTrue(d._internal_debugger.trace_counter == 223)
        d.kill()


if __name__ == "__main__":
    unittest.main()

File: test/srcs/trace_test.c
//
// This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
// Copyright (c) 2023-2024 Roberto Alessandro Bertolini. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for details.
//

#include <stdio.h>
#include <stdlib.h>

void inner(int a, int i){
    a=a+i;
}

void base_test(){
    int a=0;
    for(int i=0; i<10; i++){
        a++;
        printf("a = %d\n", a);
        inner(a,i);
    }
}

int main()
{
    base_test();

    return EXIT_SUCCESS;
}